code/modules/mob/living/simple_animal/friendly/trader.dm

```diff
@@ -94,154 +181,285 @@
-	if(npc_result != "Yes")
+	var/buyer_will_buy = show_radial_menu(user, src, npc_options, custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE, tooltips = TRUE)
+	if(buyer_will_buy != "Yes")
 		return
 	face_atom(user)
-	var/obj/item/holochip/cash
-	cash = user.is_holding_item_of_type(/obj/item/holochip)
-	if(!cash || cash.credits < products[item_to_buy])
-		say(nocashphrase)
+	if(!spend_buyer_offhand_money(user, product_info[TRADER_PRODUCT_INFO_PRICE]))
+		say(return_trader_phrase(NO_CASH_PHRASE))
 		return
-	cash.spend(products[item_to_buy])
 	item_to_buy = new item_to_buy(get_turf(user))
 	user.put_in_hands(item_to_buy)
 	playsound(src, sell_sound, 50, TRUE)
-	say(buyphrase)
+	product_info[TRADER_PRODUCT_INFO_QUANTITY] -= 1
+	say(return_trader_phrase(BUY_PHRASE))
+
+///Calculates the value of money in the hand of the buyer and spends it if it's sufficient
+/mob/living/simple_animal/hostile/retaliate/trader/proc/spend_buyer_offhand_money(mob/user, the_cost)
+	var/value = 0
+	var/obj/item/holochip/cash
+	cash = user.is_holding_item_of_type(/obj/item/holochip)
+	if(cash)
+		value += cash.credits
+	if((value >= the_cost) && cash)
+		cash.spend(the_cost)
+		return TRUE
+	return FALSE //Purchase unsuccessful
+
+/**
+ * Tries to call sell_item on one of the user's held items, if fail gives a chat message
+ *
+ * Gets both items in the user's hands, and then tries to call sell_item on them, if both fail, he gives a chat message
+ * Arguments:
+ * * user - (Mob REF) The mob trying to sell something
+ */
+/mob/living/simple_animal/hostile/retaliate/trader/proc/try_sell(mob/user)
+	var/obj/item/activehanditem = user.get_active_held_item()
+	var/obj/item/inactivehanditem = user.get_inactive_held_item()
+	if(!sell_item(user, activehanditem) && !sell_item(user, inactivehanditem))
+		say(return_trader_phrase(ITEM_REJECTED_PHRASE))
 
 /**
  * Checks if an item is in the list of wanted items and if it is after a Yes/No radial returns generate_cash with the value of the item for the NPC
  * Arguments:
- * * user - The mob trying to sell something
- * * selling - The item being sold
+ * * user - (Mob REF) The mob trying to sell something
+ * * selling - (Item REF) The item being sold
  */
-/mob/living/simple_animal/hostile/retaliate/trader/proc/sell_item(mob/user, selling)
-	var/obj/item/sellitem = selling
+/mob/living/simple_animal/hostile/retaliate/trader/proc/sell_item(mob/user, obj/item/selling)
 	var/cost
-	if(!sellitem)
+	if(!selling)
+		return FALSE
+	var/list/product_info
+	//Keep track of the typepath; rather mundane but it's required for correctly modifying the wanted_items
+	//should a product be sellable because even if it doesn't have a entry because it's a child of a parent that is present on the list
+	var/typepath_for_product_info
+	if(selling.type in wanted_items)
+		product_info = wanted_items[selling.type]
+		typepath_for_product_info = selling.type
```

Comment:
Does this work? Using just the type usually need to use ``initial()`` for it to properly work.